### PR TITLE
Ability to start test resources in parallel (concurrently)

### LINKED ISSUE #12056: Ability to start test resources in parallel (concurrently)
**Description**
At the moment it is not possible to start test resources in parallel. (`@QuarkusTestResource` annotated test with `QuarkusTestResourceLifecycleManager` implementation classes)

The use case for me, is that i'm using integration test style tests and sometimes i have to start more than one container (eg: oracle DB and keycloak auth server). It would really speed up test execution if docker could start both containers at the same time instead of one at a time.

There is currently no way of doing this, other than using a single test resource that spawns multiple containers. I want to keep using separate test resources (for reuse).

Example:
```java
@QuarkusTestResource(OracleDBTestResource.class)
@QuarkusTestResource(KeycloakTestResource.class)
public class SomeTest {

    @Test
    public void testStuff() {
(...)
```

Each test resource starts a container. OracleDBTestResource starts an oracle database and KeycloakTestResource starts keycloak.
The KeycloakTestResource#start method is only called after the OracleDBTestResource#start method completes. Which means the keycloak container is only started after the oracle container is started and ready.

What i would like is for both test resources to start in parallel. For both start methods to run simultaneously.

**Implementation ideas**
What i think would be cool from a "user" point of view is to mark the test resource as `parallel = true` (another field in the `@QuarkusTestResource` annotation). And keep it false by default for two reasons:
* for people that upgrade to keep the same behaviour as they currently have
* because if different test resources running simultaneously override the same properties (i don't really know why people would want this) when running them concurrently the last to finish will be the one actually overriding (as opposed to the last to start when running them sequentially)

**Zulip discussion that originated this issue**
https://quarkusio.zulipchat.com/#narrow/stream/187030-users/topic/test.20resources.20-.20start.20in.20parallel

### PR DESCRIPTION
**Pull request by Nunalves**
Feature #12056

### CODE REVIEW

File: test-framework/common/src/main/java/io/quarkus/test/common/TestResourceManager.java [hunk 1]
```diff
@@ -46,17 +64,41 @@ public void init() {
...
         started = true;
-        Map<String, String> ret = new HashMap<>();
-        for (TestResourceEntry entry : testResourceEntries) {
-            try {
-                Map<String, String> start = entry.getTestResource().start();
-                if (start != null) {
-                    ret.putAll(start);
+        Map<String, String> ret = new ConcurrentHashMap<>();
+
+        ExecutorService executor = newExecutor();
```
Comment (stuartwdouglas) [inline test-framework/common/src/main/java/io/quarkus/test/common/TestResourceManager.java:69]: This needs to be shutdown in a finally block

File: test-framework/common/src/main/java/io/quarkus/test/common/TestResourceManager.java [hunk 2]
```diff
@@ -69,8 +111,22 @@ public void init() {
...
+            try {
+                future.get();
+            } catch (CancellationException | InterruptedException | ExecutionException e) {
+                throw new RuntimeException("Error waiting for test resource future to finish.", e);
+            }
+        }
+    }
+
+    private ExecutorService newExecutor() {
+        return Executors.newWorkStealingPool();
```
Comment (stuartwdouglas) [inline test-framework/common/src/main/java/io/quarkus/test/common/TestResourceManager.java:125]: I would just use a fixed pool sized to the number of tasks

**Review by stuartwdouglas — CHANGES_REQUESTED** [review]

**Comment by geoand** [comment]
Thanks for this!

It seems like CI is complaining, can you have a look?

**Comment by famod** [comment]
A small addition to https://github.com/quarkusio/quarkus/blob/master/docs/src/main/asciidoc/getting-started-testing.adoc#13-starting-services-before-the-quarkus-application-starts would be helpful.

Btw, what happens if you have one non-parallel resource and one parallel one? Does the second, parallel one wait for the first to finish?

**Comment by Nunalves** [comment]
Thanks guys!

I'll implement @stuartwdouglas suggestions and check the build pipeline errors as @geoand mentioned (i left it running over night) 

> Nice feature!
> 
> A small addition to https://github.com/quarkusio/quarkus/blob/master/docs/src/main/asciidoc/getting-started-testing.adoc#13-starting-services-before-the-quarkus-application-starts would be helpful.
> 
> Btw, what happens if you have one non-parallel resource and one parallel one? Does the second, parallel one wait for the first to finish?

Thanks @famod :)
I'll add the docs.

Answering your question. All the non-parallel resources (i called it sequential resources) run in the same thread that is started along with all the parallel ones. So, in the case you mentioned, the parallel resource will not wait for the non-parallel to finish (i had to make some decisions :p)

The main reason is that you can have multiple sequential resources and you would have to wait for them all to start the first parallel one. I mean, if you wanted that behaviour, why not mark all of them as sequential anyway?

I'm missing something? Or should we leave it as is.

File: test-framework/common/src/main/java/io/quarkus/test/common/TestResourceManager.java (hunk 2 above)
Comment (Nunalves) [reply test-framework/common/src/main/java/io/quarkus/test/common/TestResourceManager.java:125]: I thought of doing that. I even searched to see if there was any shared thread pool that would be indicated to use here. I can change it, but shouldn't we have a limit at least? The first test to run will be the one starting up all the different test resources (that should never be something like 100, i don't think so, but you never know :p)

What do you think? fixed pool size to the number of tasks with a limit to the number of CPUs?

File: test-framework/common/src/main/java/io/quarkus/test/common/TestResourceManager.java (hunk 1 above)
Comment (Nunalves) [reply test-framework/common/src/main/java/io/quarkus/test/common/TestResourceManager.java:69]: Well spotted! Thanks :)
I'll fix it along with the other issue you mentioned after you respond

**Comment by famod** [comment]
@Nunalves Thanks for clarifying!
I'am fine with your approach. I was just thinking whether one would want to form "fork join groups" but we can always make things more complicated later in case it is really needed. ;-)

**Review by geoand — COMMENTED** [review]
Also keep in mind that the commits should be squashed because we use merge-commit and the UI doesn't let us squash when merging

**Comment by Nunalves** [comment]
> Also keep in mind that the commits should be squashed because we use merge-commit and the UI doesn't let us squash when merging

Thanks @geoand. I know that. I only found time to get back to this today and thought i'd apply the review suggestions and thank look into the build.

I'll probably need to do more commits to fix the failing build (after knowing what could i have broke with such a simple change, any clue would help).

But don't worry, i'll clean the mess afterwards with a squash and force push.

**Comment by geoand** [comment]
Cool, thanks

File: test-framework/common/src/main/java/io/quarkus/test/common/TestResourceManager.java (hunk 2 above)
Comment (gsmet) [reply test-framework/common/src/main/java/io/quarkus/test/common/TestResourceManager.java]: I agree limiting it to the number of CPUs is a good idea.

Comment (famod) [reply test-framework/common/src/main/java/io/quarkus/test/common/TestResourceManager.java]: This ^ plus maybe a system property to override the default?

**Comment by arvid-villen-adtoox** [comment]
Any updates on this?

This feature would save us a lot of time :)

**Comment by stuartwdouglas** [comment]
I have rebased and force pushed, but I think there were CI issues previously.

**Comment by arvid-villen-adtoox** [comment]
@geoand here is shameless plug for this issue. Is there any chance that you could look at/finish this?
I totally understand if not! And please tell me if I'm out of line in pinging you!

**Comment by geoand** [comment]
Not out of line at all :)

Unfortunately I don't see myself having time to look at this any time soon.

**Comment by arvid-villen-adtoox** [comment]
Ok, thanks for the fast reply!

**Comment by famod** [comment]
I'll try to find out what goes wrong. For that I rebased again and added an exception handling fix to `QuarkusTestExtension#getAdditionalTestResources`.
Now let's see what CI spits out...

**Comment by stuartwdouglas** [comment]
I have pushed a fix, lets see what CI says now.

**Comment by arvid-villen-adtoox** [comment]
@stuartwdouglas @famod Thanks a lot fixing this issue!
As CI is passing we only need to merge?

**Review by stuartwdouglas — APPROVED** [review]

**Comment by Nunalves** [comment]
Coming here after some time! Thanks @famod @stuartwdouglas for fixing CI
I'm sorry that my first contribution to quarkus was an incomplete one. Lets see if i find time to compensate you guys :p
Still, glad that i contributed with something :p

**Comment by famod** [comment]
@Nunalves no worries at all! Stuart and I didn't have to do much.
Again thanks and keep the contributions coming!  😁

**Pull request merged**